### ISSUE TITLE
Feature request: Sorting points to minimize dome slewing

### ISSUE THREAD

**Issue by ACMBakx** [mworion/MountWizzard4#196]
[Model_Sorted_EW.xlsx](https://github.com/mworion/MountWizzard4/files/7324765/Model_Sorted_EW.xlsx)
[Model_Sorted_for_Dome.xlsx](https://github.com/mworion/MountWizzard4/files/7324766/Model_Sorted_for_Dome.xlsx)
![Model59pts](https://user-images.githubusercontent.com/59890758/136842775-bdfbd19a-fd91-43a0-a5c1-c91a443bf29a.JPG)
I just want to bring up the idea once more.
Attached 2 sheets. I copied the numbers from a log file of a real run I did recently.
One is showing how MW4 sorts the points in E/W order as is shown in the image.
You see how the dome slews forward **and backwards** to follow points that run from East to West.
The other one is rearranged and shows the sequence for minimal dome slews.
Would it be useful to have this "sorting option" too?

**Comment by ACMBakx** [comment]
SUCCESS!!
We have some kind of clear sky for a few hours, so just ran the point model again.
This time every point was platesolved and MW4 performed as it should.
As to be predicted, around 180° Az the mount was flipping like a madman.
Pier side East. Pier side West. And back again.
But the dome was steady. Multiple images were taken without slewing the dome.
Duration of the run was just under 30 mins. You can tell from the log. (27 mins)
So the duration was about the same as the standard East/West sorting run.
A lot of time was spend on the numerous meridian flips. Some 10-15 I believe.
But the dome slewing (or lack there of) was significant reduced.
PAE and RMS are the same as before. (after knocking off one point)
So for me this run was succesfull.

And I have made my point . . . . . . . .   :-)

Arie




[mw4-2021-10-15.log](https://github.com/mworion/MountWizzard4/files/7355732/mw4-2021-10-15.log)
t.

**Comment by mworion** [comment]
Hi Arie,

Good to hear! Would you accept such a solution for sorting, regardless of the flips the mount does? 

Michel

**Comment by ACMBakx** [comment]
I would welcome it for sure.
But if I am the only user with his request, do not bother.
You'll have to calculate all the dome positions up front obviously.
But It wil be up to the user to determine what is most feasable for him.
I can imagine if you have a big mount with multiple heavy telescopes on it and/or are having flex/flop problems, that much flipping may not be welcome. But a big and heavy dome could greatly benefit from the lesser amount of slewing.
As I said before, it will be a trade-of between the two.
My mount is light (1000hps) and carries only one light and stiff telescope. I have the slew rate set at max 7° .
But a 3000 or 4000hps loaded to its max will be a different situation.
Maybe check with Nicolás what he thinks about the idea.

Arie

**Comment by InFINNity-Deck** [comment]
Hi Arie and Michel,

In my case the dome slews at about the same speed as the mount, so there is little gain in having one or the other doing the slews. The only major improvement would be avoiding a 180 degrees dome slew for a minor mount slew.
Still I find this an interesting concept, but fear for two things:
- the amount of additional stress put on a mount when fully loaded
- quality degradation due to flexure

The first is only marginally worrying, adding some more heavy slews to the mount's life. After all we are not doing this every night... ;-) If it would become an issue, the mount slew speed could be lowered. 

The flexure is another story. Any flexure in the mounting of the scope, it's focuser or camera mounting could be adding to inaccuracies, depending on the speed at which this flexure happens. The are quite a few people around imaging with SCTs, which have the habit to suffer from mirror-flop due to play between the mirror-tube and central-baffle-tube. The grease in between the two is generally quite thick and I have seen mirror-flop occurring minutes after the scope was flipped. Passing the meridian with a single flip will produce one odd image, while passing it over and over again may result in either all good images or all bad, depending on the time between images and the time required for the grease to let go of the mirror.

Nicolàs

**Comment by ACMBakx** [comment]
I have to add to the conversation that in my configuration with a small dome (they don’t come any smaller) the effect of the GEM distance is quite severe. It causes a wide area Az were the mount will peek around the corner in two directions, causing lots of flips. I was expecting a comment like Nicolás is making, but needed to here from himself. Thank you Nicolás for responding.
Also in my case there was no gain in time. What was gained by one, was lost in the other.
I have my mount settling time set at 2 sec only. (Yet the RMS=5) But some have it set at 20sec(?) with a reduced slew speed. 
Maybe you can put it out in a beta version and let others try and comment on it. If it is not liked, you throw it out again.
I will keep the dome sorted points-file for myself anyway.

Arie

**Comment by mworion** [comment]
Hi Arie,
first having your log file with the break after the 14 point inspected, I see a new model start after point 14. So this made mixing a second model run in a already running one. Do you have any idea what happened ? The result is obvious as MW4 could not handle two sessions in parallel - > does not make sense. I try t avoid this, but the might be a possible user input or technical backdoor which enables a second run.

To the dome topic:
Sorting the azimuth as real azimuth of the dome should be possible. The point with the remarks of Nicolas is, is that in overall a new sorting offer or do we (I) have to think about further optimizations to make that usable on a broader base.
Michel

**Comment by ACMBakx** [comment]
When MW4 stopped at station 14, I sat and waited a while for things to re-commence. When that didn’t happen, I clicked the Run button again and MW4 appeared to restart all over.
I should have stopped and restarted MW4, I suppose, to get a clean start again.
A user-input error I would say.
I am not sure that I understand what you mean with “usable on a broader base”. 
Can you offer it as a 3rd option, which is only shown/offered when in the settings a dome is configured?
Then non-dome users are not bothered/confused by it.

**Comment by mworion** [comment]
Hi Arie, Nicolas,
though about this my proposal: I do a regular points cloud. This could also be used with features like auto delete points within meridian limits (reducing flips there). Then calculating all real dome Azimuth values and sorting the point cloud for East / West but now for dome Az and not for mount Az. So I get the following results for Aries demofile

Original manual sorting from Arie:
![Uploading Bildschirmfoto 2021-10-16 um 17.57.35.png…]()

Using that data for East / West sorting: 
![Bildschirmfoto 2021-10-16 um 17 58 30](https://user-images.githubusercontent.com/6632980/137594165-ad87793c-6c25-4546-9c49-78789acab9fa.png)
I agree, that will move the dome more frequently

Now adding the dome Az sorting:
![Bildschirmfoto 2021-10-16 um 18 00 23](https://user-images.githubusercontent.com/6632980/137594203-c27048d8-f361-43b5-94ec-d9e0fa0c95ca.png)
This is not far away from the manual sorting. But as said could be used for any selected method of getting a point cloud.

For avoiding to much flip, you could add the constraint with meridian limits and you get:
![Bildschirmfoto 2021-10-16 um 18 03 03](https://user-images.githubusercontent.com/6632980/137594282-2722cf80-1778-434a-8e99-074c954e8bb2.png)

So this work is done. Arie: could you check in you manual sorting with the result of the automatic one (3rd). I know it's the other way round for the dome, but should not introduce more slews. If that makes sense, I would ask you to test this, I will build a beta in an hour (2.1.0b7) with the feature implemented. Please be aware that calculating the dome Az for all the build points through the mount (I need to know which flip side it will take) take some time. You will see the progress bar in sorting group:

![Bildschirmfoto 2021-10-16 um 18 06 34](https://user-images.githubusercontent.com/6632980/137594391-63a61e61-295f-4310-a6f1-646aa8e582f8.png)

Feedback welcome.
Michel

**Comment by ACMBakx** [comment]
Upgradet to 2.1.0b7 successfully.
But now MW4 does not start anymore.
I see in a blink in the cmd-window a few lines flashing by, Then this window closes again.
No log is being generated.

**Comment by mworion** [comment]
Hi Arie,
unfortunately I cannot see any point with this behavior on my windows10 test equipment. 
From which version did you upgrade? For an analyse could you post your config file?
To restart make a backup your config directory and empty it after that, please try to start MW4 again.
Michel

**Comment by ACMBakx** [comment]
They are both part of the same event.
I managed to stop the movieplayer a fraction of a second apart and took screenshots. 
The last line says there is a type error somewhere . . . . . .?
I can easily go back installing 2.0.5 or 6 using the command line pip install mountwizzard4==etc.
Tried to do that with 2.1.0b7. Get the message that the installation went ok.
But it does not start up.

**Comment by mworion** [comment]
Arie,
Just make a new folder and copy the scripts in the folder and run install. You get a second, fully independent install. Nothings else to change, please keep everything else. To run MW4 in new install, run MW4Run from the new folder.
Michel

Von meinem iPad gesendet

**Comment by mworion** [comment]
Hi Arie,
still working robustness, but still can't catch the issue in my setup. I set this point as remark.
If you would like, have a look to the (hopefully) last beta before the release 2.1. It also adds some color themes:
![red_night_theme](https://user-images.githubusercontent.com/6632980/139580078-6a5f66b6-c007-4937-a29c-95bff06a8125.png)
![light_day_theme](https://user-images.githubusercontent.com/6632980/139580089-e745383c-3ac5-4316-9a96-e70ea5b3fa55.png)
I close the original issue as sorting work out nicely.
Michel

**Comment by ACMBakx** [comment]
OK, wildo AFO.
Color schemes will be appreciated I am sure.
Arie

Verstuurd vanaf mijn iPad
